src/librustc_driver/driver.rs

```diff
@@ -826,6 +828,391 @@ pub fn phase_2_configure_and_expand<F>(sess: &Session,
     })
 }
 
+/// Trace of Queries
+///
+/// Formally, a trace of the queries consists of a tree, where
+/// sub-trees represent sub-traces. In particular, the nesting
+/// structure of the trace of queries describes how the queries depend
+/// on one another.
+///
+/// Even more precisely, this tree represents a directed acyclic graph
+/// (DAG), where shared sub-graphs consist of tree nodes that occur
+/// multiple times in the tree, first as "cache misses" and later as
+/// "cache hits".
+///
+/// See also:
+/// https://github.com/matthewhammer/rust-forge/blob/master/profile-queries.md#trace-of-queries
+mod trace {
```

Comment:
I would not put this code here. I'd prefer a separate file. We could make `driver/mod.rs` and put it in `driver/trace.rs` (`M-x rust-promote-module-into-dir`, in emacs), or else just make it a sibling.